### ISSUE TITLE
[Bug]: Logits processing with Lora is incorrect

### ISSUE THREAD

**Issue by yinghai** [vllm-project/vllm#17766]
### Your current environment

V1. Logic error. 

### 🐛 Describe the bug

When we have mixed prompt and sampling in a batch (possible in V1), we can no longer use indices_padded = self.punica_wrapper.sampler_indices_padded to select from lora_logits of size [(lora_max+1*B, d_model]

https://github.com/vllm-project/vllm/blob/8d84d836d17bdabe4c640021bc6f8bd11a546a44/vllm/lora/layers.py#L1142

But indices_padded is of size samples which is <= B. Hence the result would be incorrect or error out later in
https://github.com/vllm-project/vllm/blob/8d84d836d17bdabe4c640021bc6f8bd11a546a44/vllm/lora/layers.py#L1152
due to broadcast mismatch.

To fix, we probably need to add lora mapping for  each token in the batch, say, batch_indices like 
https://github.com/vllm-project/vllm/blob/ba7703e659e21ea376c25b872c4a80a7c82804fe/vllm/lora/punica_wrapper/utils.py#L130
And the 2D flat indexing like https://github.com/iterationlab/monorepo/blob/7345300cc615de19fb54df261447c1356c8dd028/third_party/vllm/vllm/lora/punica_wrapper/utils.py#L133

And use that instead of `sampler_indices_padded` to indice the lora_logits. 

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by jeejeelee** [comment]
I will look at this ASAP.  @varun-sundar-rabindranath  If you have a chance, would you please take a look at this as well?

**Comment by yinghai** [comment]
Thanks! I think maybe this is more of a mismatch of logits batch size when there is prompt logprob request. Because then you need to put prompt tokens through logits processor. For example if I have a request with 5 prompt tokens and ask for prompt logprob, the logit processor batch size will be 4 but sampler indices size will just be 1, hence the mismatch.

**Comment by jeejeelee** [comment]
Yes, there is a bug in V1 lora's prompt_logprob. You can refer to: https://github.com/vllm-project/vllm/pull/16694
